proto/prost/private/prost.bzl

```diff
@@ -0,0 +1,435 @@
+            "_process_wrapper": attr.label(
+                doc = "A process wrapper for running rustc on all platforms.",
+                default = Label("//util/process_wrapper"),
+                executable = True,
+                allow_single_file = True,
+                cfg = "exec",
+            ),
+        },
+        fragments = ["cpp"],
+        host_fragments = ["cpp"],
+        toolchains = [
+            "@bazel_tools//tools/cpp:toolchain_type",
+            "@rules_rust//proto/prost:toolchain_type",
+            "@rules_rust//rust:toolchain_type",
+            "@rules_rust//rust/rustfmt:toolchain_type",
+        ],
+        incompatible_use_toolchain_transition = True,
+    )
+
+_rust_prost_aspect = _make_rust_prost_aspect(
+    doc = "An aspect used to generate and compile proto files with Prost.",
+    is_tonic = False,
+)
+
+_rust_tonic_aspect = _make_rust_prost_aspect(
+    doc = "An aspect used to generate and compile proto files with Prost and Tonic.",
+    is_tonic = True,
+)
+
+def _rust_prost_library_impl(ctx):
+    proto_info_provider = TonicProtoInfo if ctx.attr._is_tonic else ProstProtoInfo
+
+    proto_dep = ctx.attr.proto
+    rust_proto_info = proto_dep[proto_info_provider]
+    dep_variant_info = rust_proto_info.dep_variant_info
+
+    return [
+        DefaultInfo(files = depset([dep_variant_info.crate_info.output])),
+        rust_common.crate_group_info(
+            dep_variant_infos = depset(
+                [dep_variant_info],
+                transitive = [rust_proto_info.transitive_dep_infos],
+            ),
+        ),
+    ]
+
+def _make_rust_prost_library_rule(doc, is_tonic):
+    return rule(
+        doc = doc,
+        implementation = _rust_prost_library_impl,
+        attrs = {
+            "proto": attr.label(
+                doc = "A `proto_library` target for which to generate Rust gencode.",
+                providers = [ProtoInfo],
+                aspects = [_rust_tonic_aspect if is_tonic else _rust_prost_aspect],
+                mandatory = True,
+            ),
+            "_collect_cc_coverage": attr.label(
+                default = Label("@rules_rust//util:collect_coverage"),
+                executable = True,
+                cfg = "exec",
+            ),
+            "_is_tonic": attr.bool(
+                doc = "Indicates whether or not Tonic behavior should be enabled.",
+                default = is_tonic,
+            ),
+        },
+    )
+
+rust_tonic_library = _make_rust_prost_library_rule(
+    doc = "A rule for generating a Rust library using Prost and Tonic.",
+    is_tonic = True,
+)
+
+rust_prost_library = _make_rust_prost_library_rule(
+    doc = "A rule for generating a Rust library using Prost.",
+    is_tonic = False,
+)
+
+def _rust_prost_toolchain_impl(ctx):
+    is_tonic = ctx.attr.tonic_runtime != None
+
+    proto_lang_toolchain = proto_common.ProtoLangToolchainInfo(
+        out_replacement_format_flag = "--tonic_out=%s" if is_tonic else "--prost_out=%s",
```

Comment:
This seems kinda wrong, shouldn't this always be `--prost_out=%s` in the event a user has a toolchain with tonic on it but only wants prost outputs (e.g. just uses `rust_prost_library`)?